core/src/main/kotlin/xtdb/arrow/DecimalVector.kt

```diff
@@ -0,0 +1,52 @@
+package xtdb.arrow
+
+import org.apache.arrow.memory.BufferAllocator
+import org.apache.arrow.vector.types.pojo.ArrowType
+import xtdb.api.query.IKeyFn
+import xtdb.arrow.metadata.MetadataFlavour
+import xtdb.util.Hasher
+import java.math.BigDecimal
+
+class DecimalVector private constructor(
+    override var name: String, override var nullable: Boolean, override var valueCount: Int,
+    override val validityBuffer: ExtensibleBuffer, override val dataBuffer: ExtensibleBuffer,
+    val precision: Int, val scale: Int, val bitWidth: Int,
+) : FixedWidthVector(), MetadataFlavour.Number {
+
+    private val BIT_WIDTHS = setOf(32, 64, 128, 256)
+
+    init {
+       require(bitWidth in BIT_WIDTHS) { "Invalid bit width for DecimalVector: $bitWidth" }
+    }
+
+    constructor(al: BufferAllocator, name: String, nullable: Boolean, precision: Int, scale: Int, bitWidth: Int)
+            : this(name, nullable, 0, ExtensibleBuffer(al), ExtensibleBuffer(al), precision, scale, bitWidth)
+
+    override val byteWidth = (bitWidth / 8)
+
+    override val type: ArrowType = ArrowType.Decimal(precision, scale, bitWidth)
+
+    override fun getObject0(idx: Int, keyFn: IKeyFn<*>) : BigDecimal =
+       dataBuffer.readBigDecimal(idx, scale, byteWidth)
+
+    override fun writeObject0(value: Any) {
+        if (value is BigDecimal) {
+            if (value.precision() > precision || value.scale() != scale) {
+                throw InvalidWriteObjectException(fieldType, value)
+            }
+            try {
+                dataBuffer.writeBigDecimal(value, byteWidth)
+            } catch (e: UnsupportedOperationException) {
```

Comment:
why might this throw UOE, sorry?

Reply:
If your decimal doesn't fit your byteWidth.

Reply:
ah, kk - that's one for IllegalArgumentException - UOE is something the code doesn't support (yet); IAE is 'caller messed up'

Reply:
I'd stick that as a `require` in `writeBigDecimal` - that throws IAE if the condition isn't met

Reply:
I think this will likely never happen if we properly deal with checking the precision before. It's the `DecimalUtility` that throws this error. I think it can only happen if you create a arrow decimal type where the precision can't fit the bitwidth, but I guess the assumption here is that you then messed up elsewhere.